pkg/compactor/block_upload.go

```diff
@@ -354,13 +400,79 @@ func (c *MultitenantCompactor) uploadMeta(ctx context.Context, logger log.Logger
 	blockID ulid.ULID, name string, userBkt objstore.Bucket) error {
 	dst := path.Join(blockID.String(), name)
 	level.Debug(logger).Log("msg", fmt.Sprintf("uploading %s to bucket", name), "dst", dst)
-	buf := bytes.NewBuffer(nil)
-	if err := json.NewEncoder(buf).Encode(meta); err != nil {
-		return errors.Wrap(err, "failed to encode block metadata")
-	}
-	if err := userBkt.Upload(ctx, dst, buf); err != nil {
+	if err := marshalAndUploadToBucket(ctx, userBkt, dst, meta); err != nil {
 		return errors.Wrapf(err, "failed uploading %s to bucket", name)
 	}
+	return nil
+}
+
+func (c *MultitenantCompactor) validateBlock(ctx context.Context, blockID ulid.ULID,
+	userBkt objstore.Bucket, meta metadata.Meta) error {
+	blockDir, err := os.MkdirTemp(filepath.Join(c.compactorCfg.DataDir), "upload")
+	if err != nil {
+		return errors.Wrap(err, "failed to create temp dir")
+	}
+	defer func() {
+		if err := os.RemoveAll(blockDir); err != nil {
+			level.Warn(c.logger).Log("msg", "failed to remove temp dir", "path", blockDir, "err", err)
+		}
+	}()
+
+	if err := userBkt.Iter(ctx, blockID.String(), func(pth string) error {
+		fname := filepath.Join(blockDir, filepath.FromSlash(pth))
+		if strings.HasSuffix("/", fname) {
+			if err := os.Mkdir(fname, os.ModeDir); err != nil {
```

Comment:
Isn't it safer to use `os.MkdirAll`, in case you get a nested directory, where the parent doesn't exist? I don't see it doing any harm.
```suggestion
			if err := os.MkdirAll(fname, os.ModeDir); err != nil {
```

Reply:
Are you also guaranteed to get each file's directory from the iterator? Maybe it's better to create each file's directory (with `os.MkdirAll`) if necessary, to be robust?

Reply:
There are no extra directories to be created.

> Are you also guaranteed to get each file's directory from the iterator?

Yes, iterator returns objects in sorted order. It returns "directories" before any objects in it.

Reply:
What made me think that you might not receive directories, is that I'm used to Google Storage f.ex. not having a concept of directories. It just has object names IIRC.

Reply:
Oh, I think you're right: when using recursive listing, I don't think we will get the directories as separate entries.

But this code also doesn't *ask* for recursive listing, so it won't even find the chunks. (That's a bug.)

Reply:
(Btw, GCS would return "directories", or synthetic prefixes, if you set Delimeter option of `Objects` call.)

Reply:
> (Btw, GCS would return "directories", or synthetic prefixes, if you set Delimeter option of Objects call.)

TIL.